### PR TITLE
[WIP] Rework parsing Cabal metadata

### PR DESCRIPTION
**Pull request by snowleopard**
The current implementation of parsing Cabal data is a big mess, which affects performance as described in #671 and #550. In this PR I intend to simplify the implementation and avoid unnecessary reparsing.

This is in a very draft state right now and not ready for review. I'm pushing to do a CI run.

### CODE REVIEW

**Comment by alpmestan** [comment]
Since this is what we're after here, do you think you could quantify the improvement that this patch makes, **once it's ready for review** (i.e: no rush, I understand it's not quite ready yet) ? Perhaps give the timings for no-op runs? If you don't mind.

**Comment by snowleopard** [comment]
> do you think you could quantify the improvement that this patch makes, 

@alpmestan Yes, absolutely! I'll try to quantify the impact both on the full build and on the zero build. Presumably the impact on the full build will be negligible, and I hope to get back to a few seconds for the zero build (compared to current 30 seconds).

I'll aim to finalise this PR by end of today.

**Comment by snowleopard** [comment]
@alpmestan Current status: zero build down to 10 seconds. Changes to full build time seem insignificant.

There is still some work to be done though.

**Comment by alpmestan** [comment]
@snowleopard Ah, that's already much more reasonable, good work! I certainly won't stand in your way if you want to reduce this further. :)

**Comment by alpmestan** [comment]
We certainly didn't see those on my last few PRs, did we? In which case it'd quite likely be related to commits recently merged to HEAD.

Any chance https://github.com/ghc/ghc/commit/ce3897ffd6e7c8b8f36b8e920168bac8c7f836ae is causing them?

**Comment by snowleopard** [comment]
> any chance ghc/ghc@ce3897f is causing them?

This was my first guess too, but it doesn't seem to touch any sensitive code -- it essentially fixes a bug in the configure scripts, and in the worst case the build system would fail during the configuration.

**Comment by alpmestan** [comment]
Any chance this fixes your problem: https://phabricator.haskell.org/D5163 ?

**Comment by snowleopard** [comment]
@alpmestan Aha, that's it! Thank you.

**Comment by snowleopard** [comment]
P.S.: Sorry, got distracted from finishing this PR at work. Will be back in action tomorrow.

**Comment by snowleopard** [comment]
Got zero build down to 8 seconds but at the cost of one extra oracle. Looks a bit messy...

**Comment by alpmestan** [comment]
Well then you're apparently saving enough time to get us below the max duration :-)

I'll quite likely review this on monday. Can't wait to see it in action.

File: src/Hadrian/Oracles/TextFile.hs [hunk 1]
```diff
@@ -73,12 +72,37 @@ lookupDependencies depFile file = do
...
+    deriving (Binary, Eq, Hashable, NFData, Show, Typeable)
+type instance RuleResult KeyValue = Maybe String
+
+newtype KeyValues = KeyValues (FilePath, String)
+    deriving (Binary, Eq, Hashable, NFData, Show, Typeable)
+type instance RuleResult KeyValues = Maybe [String]
+
+-- | This oracle reads and parses text files to answer various queries, caching
+-- and tracking the results.
+textFileOracle :: Rules ()
```
Comment (alpmestan) [inline src/Hadrian/Oracles/TextFile.hs]: Same as for the other caches, it'd be nice to get a sliiiightly more detailed overview. "Text files" just seems a little too vague. It'd probably be a good idea to mention which text files this actually gets used on. (.cabal files I suppose, but maybe others?)

**Comment by snowleopard** [comment]
@alpmestan Many thanks for the review -- I fully agree about comments. Will address all your comments tomorrow!

File: src/Hadrian/Oracles/Cabal/Type.hs [hunk 2]
```diff
@@ -23,10 +23,22 @@ import Hadrian.Haskell.Cabal.Type
 import Hadrian.Package
 import Stage
 
+-- | This type of oracle key is used by 'Hadrian.Oracles.Cabal.readPackageData'
+-- to cache reading and parseing of 'Hadrian.Haskell.Cabal.Type.PackageData'.
```
Comment (alpmestan) [inline src/Hadrian/Oracles/Cabal/Type.hs]: parsing

File: src/Hadrian/Oracles/Cabal/Type.hs [hunk 3]
```diff
@@ -23,10 +23,22 @@ import Hadrian.Haskell.Cabal.Type
...
 newtype PackageDataKey = PackageDataKey Package
     deriving (Binary, Eq, Hashable, NFData, Show, Typeable)
 type instance RuleResult PackageDataKey = PackageData
 
+-- | This type of oracle key is used by 'Hadrian.Oracles.Cabal.readContextData'
+-- to cache reading and parseing of 'Hadrian.Haskell.Cabal.Type.ContextData'.
```
Comment (alpmestan) [inline src/Hadrian/Oracles/Cabal/Type.hs]: parsing

Comment (snowleopard) [reply src/Hadrian/Oracles/Cabal/Type.hs]: Thanks, fixed both typos.

**Comment by snowleopard** [comment]
@alpmestan Thank you for the review! Merged.

**Pull request merged**